designs/2019-typescript-eslintrc-support/README.md

```diff
@@ -0,0 +1,351 @@
+- Start Date: 2019-11-18
+- RFC PR: https://github.com/eslint/rfcs/pull/50
+- Authors: Gareth Jones ([@g-rath](https://github.com/g-rath))
+
+# First-Class Support For `.eslintrc.ts`
+
+## Summary
+
+Support the ability to consume `.eslintrc.ts` configuration files natively.
+
+## Motivation
+
+Currently configuration settings are validated by JSON Schema, which while
+powerful does not have ubiquitous & high quality support across editors &
+tooling.
+
+It also tends to generate somewhat cryptic messages, especially when using
+complex elements such as combining `oneOf`, `anyOf`, and other such schemas.
+
+Meanwhile, TypeScript is a powerful & popular superset of Javascript that has
+high quality support as a first class language in the majority of editors, and
+is quickly gaining ubiquitous support in general.
+
+By allowing first-class support for `.eslintrc.ts`, the amount of configuration
+problems & general confusion over such experienced by users can be reduced,
+since it would enable us to provide tooling to type configuration files and so
+allow TypeScript to catch ESLint configuration errors at compile time, rather
+than run time.
+
+While we can already hack ways to use `.eslintrc.ts`, none of them lend
+themselves to ease-of-use in the same manner that ESLint encourages, and so
+first-class support for `.eslintrc.ts` will go long way towards encouraging
+community driven typing of ESLint plugins.
+
+Finally, ESLint is being adopted as the linter for TypeScript, meaning
+first-class support for `.eslintrc.ts` will help enable a positive feedback loop
+to help improve the two tools.
+
+There are some potential UX benefits for users that use & understand TypeScript,
+in the form
+
+## Detailed Design
+
+- Adds support for `.eslintrc.ts`
+- Requires uses to install `ts-node` manually relative to the `node_modules` of
+  the configuration file, per standard Node
+  [module resolution logic](https://nodejs.org/api/modules.html#modules_loading_from_node_modules_folders)
```

Comment:
OK, this may be the last concern.

What does it behave if multiple `ts-node`s were found? For example:

```
<root>
├ node_modules/
│ └ ts-node/
├ src/
│ └ index.ts
├ test/
│ ├ node_modules/
│ │ └ ts-node/
│ ├ index.ts
│ └ .eslintrc.ts
└ .eslintrc.ts
```

When ESLint lints `test/index.ts`, it loads both `test/.eslintrc.ts` and `.eslintrc.ts` then merges those. I'm not sure how two `ts-node`s work in that case. (this is a rare case, but it may happen on monorepo-style repositories)

In other words, two `.eslintrc.ts` files may depend different versions of TypeScript. I'm guessing that we should throw an error in such a case...

Reply:
`ts-node` doesn't actually require TypeScript - it has it as a peerDependency.

> this is a rare case, but it may happen on monorepo-style repositories

It's actually *very* rare - in a monorepo, you still should have a top level dependency on TypeScript, as it uses project composition to link the libraries together, and generally share cache, even if only one of the projects has TypeScript.

Your question is an interesting one, but I think it's a none issue, as it has the same answer to "What if you have two different versions of Javascript? i.e ES3 & ES6".

The complexity of trying to figure out if you're consuming two different versions of TypeScript across folders would be a tough one I think, especially vs just letting TypeScript fail if there's a version different.

Currently here's what will happen:

```
<root>
├ node_modules/
│ └ typescript/ <-- this version is used for (A) & (C)
├ src/
│ ├ index.ts
│ └ .eslintrc.ts (A)
├ test/
│ ├ node_modules/
│ │ └ typescript/ // <-- this version is used for (B)
│ ├ index.ts
│ └ .eslintrc.ts (B)
└ .eslintrc.ts (C)
```

Node module resolution really does take care of everything for us in this case :)

Reply:
@mysticatea this has actually got me a bit worried - I know RFCs are about RFCing and not actual implementation, but I'm just worried about the complexity supporting this would add.

I've tested it out, and it seems that b/c of the way requiring configs work, it'll always take the TypeScript from the top-level node modules.

Personally, I don't think this is a bad thing - Generally it's not a good idea to mix and match your TypeScript versions, and that will only apply to `.eslintrc.ts`, *not* the rest of your project, if you happened to have a tree like the one I described.

Reply:
In fact, my concern came from the implementation. Because the `register()` function of `ts-node` modifies a global thing, I worried if it can work or not if it does `register()` twice. If it works, I'm fine with it. If the twice `register()` call causes cryptic errors, I thought better if we make an error before doing `register()` twice.

Reply:
Ah in that case we should be in the clear, as that's why we call `enabled(false)` - `ts-node` has some issues relating to that in the past, and that was implemented to let you just disable the registerer to make it a no-op.

In addition, `ts-node` exposes an instance on `process` via a Symbol that can be used to check for its existence, and then finally typically you shouldn't be importing anything in on the path of `eslint` that would require `ts-node`, due to the nature of ESLint itself.

All in all, I believe we're safe on this front, and feel confident that there are enough potential solutions that could be implemented to consider this an acceptable risk.

Reply:
Thank you for the clarification! My concern has been resolved.